src/PluginClientBuilder.php

```diff
@@ -0,0 +1,77 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Http\Client\Common;
+
+use Http\Client\HttpClient;
+use Http\Client\HttpAsyncClient;
+use Psr\Http\Client\ClientInterface;
+
+/**
+ * Build an instance of a PluginClient with a dynamic list of plugins.
+ *
+ * This builder is immutable, each "build" methods will instead create a new
+ * instance of itself with or without the affected plugin, destroying a
+ * potentially http client built in the new instance.
+ *
+ * @author Baptiste Clavié <clavie.b@gmail.com>
+ */
+final class PluginClientBuilder
+{
+    /** @var Plugin[] */
+    private $plugins = [];
+
+    /** @var ClientInterface | HttpClient | HttpAsyncClient */
+    private $client;
+
+    /** @var ?PluginClient */
+    private $pluginClient;
+
+    /**
+     * @param ClientInterface | HttpClient | HttpAsyncClient $client Client to nest into the plugin client
+     */
+    public function __construct($client)
+    {
+        if (!$client instanceof ClientInterface && !$client instanceof HttpClient && !$client instanceof HttpAsyncClient) {
```

Comment:
If I understood @Nyholm, this is not specially for 2.0 (at least not blocking for it). I didn't check if there was a 1.x branch.

I'm not sure either if we should check for the `ClientInterface` though...

Reply:
- Master branch is v2
- Yes there is a branch for v1...
- Master branch which is v2 has as min requirement httplug: ^2.0 . The client on this version extends `ClientInterface`

> I'm not sure either if we should check for the ClientInterface though...

- Therefore my suggestion was to check only ClientInterface...

:)

Reply:
Checking all three would make more sense for v1.

Reply:
I'll retarget if there's hope for a 1.x release then. @Nyholm @sagikazarmark ?

Reply:
> Therefore my suggestion was to check only ClientInterface...

ofc you must keep the check for the Async client..

Reply:
why aren't you sure about that ?

Reply:
I removed the `HttpClient`, even though I'm also not sure due to @dbu's comment... I saw in the PR that introduced http-plug 2.0 (psr-18 compatible) that removing the compatibility for 1.0 was on purpose because it involved some magic.

So I should either target 1.x and thus support also the `HttpClient`, or leave it as is and remove it in here. But, AFAIC, new features should only target 2.0 (master), so....